direct/io-cassandra/src/main/java/cz/o2/proxima/direct/cassandra/CassandraLogReader.java

```diff
@@ -69,68 +71,91 @@
-                Row row = rowIter.next();
-                String key = row.getString(0);
-                int field = 1;
-                for (AttributeDescriptor<?> attribute : attributes) {
-                  String attributeName = attribute.getName();
-                  if (attribute.isWildcard()) {
-                    // FIXME: this is wrong
-                    // need mapping between attribute and accessor
-                    String suffix = accessor.getConverter().asString(row.getObject(field++));
-                    attributeName = attribute.toAttributePrefix() + suffix;
-                  }
-                  ByteBuffer bytes = row.getBytes(field++);
-                  if (bytes != null) {
-                    byte[] array = bytes.slice().array();
-                    if (!observer.onNext(
-                        StreamElement.upsert(
-                            accessor.getEntityDescriptor(),
-                            attribute,
-                            "cql-"
-                                + accessor.getEntityDescriptor().getName()
-                                + "-part"
-                                + p.getId()
-                                + position.incrementAndGet(),
-                            key,
-                            attributeName,
-                            System.currentTimeMillis(),
-                            array),
-                        BatchLogObservers.defaultContext(p))) {
-
-                      cont = false;
-                      break;
-                    }
-                  }
-                }
+            for (Partition p : partitions) {
+              if (!processSinglePartition(
+                  (CassandraPartition) p, attributes, terminationContext, observer)) {
+                break;
               }
             }
-            observer.onCompleted();
+            terminationContext.finished();
           } catch (Throwable err) {
-            observer.onError(err);
+            terminationContext.handleErrorCaught(
+                err, () -> observeInternal(partitions, attributes, observer, terminationContext));
           }
         });
   }
 
+  private boolean processSinglePartition(
+      CassandraPartition partition,
+      List<AttributeDescriptor<?>> attributes,
+      TerminationContext terminationContext,
+      BatchLogObserver observer) {
+
+    ResultSet result;
+    Session session = accessor.ensureSession();
+    result =
+        accessor.execute(accessor.getCqlFactory().scanPartition(attributes, partition, session));
+    AtomicLong position = new AtomicLong();
+    for (Row row : result) {
+      if (terminationContext.isCancelled()) {
+        return false;
+      }
+      String key = row.getString(0);
+      int field = 1;
+      for (AttributeDescriptor<?> attribute : attributes) {
+        String attributeName = attribute.getName();
+        if (attribute.isWildcard()) {
+          // FIXME: this is wrong
```

Comment:
do we have an issue for this?

Reply:
I'd say we do not. This is not modified in this PR.